### PR TITLE
CORE-7269 metadata add component group structure and mgm group param hash

### PR DESCRIPTION
**Pull request by vlajos**
* CORE-7269 Add component group structure to Transaction Metadata. Also calculate the number of component group from this, instead of storing that explicitly.
* CORE-7269 Store Group Parameters hash in Transaction Metadata.
* CORE-7269 Persist group Parameters when starting to use their hash in tx metadata.
* Load back stored group parameters for later verification.
* Added a kryo serializer for SignedGroupParameters.
* Added a new GroupParametersFactory create().

API: https://github.com/corda/corda-api/pull/1042

### CODE REVIEW

File: components/ledger/ledger-common-flow/src/main/kotlin/net/corda/ledger/common/flow/impl/transaction/factory/TransactionMetadataFactoryImpl.kt [hunk 1]
```diff
@@ -24,19 +25,32 @@ class TransactionMetadataFactoryImpl @Activate constructor(
...
             TransactionMetadataImpl.CPI_METADATA_KEY to getCpiSummary(),
             TransactionMetadataImpl.CPK_METADATA_KEY to getCpkSummaries(),
-            TransactionMetadataImpl.SCHEMA_VERSION_KEY to TransactionMetadataImpl.SCHEMA_VERSION
+            TransactionMetadataImpl.SCHEMA_VERSION_KEY to TransactionMetadataImpl.SCHEMA_VERSION,
+            TransactionMetadataImpl.MEMBERSHIP_GROUP_PARAMETERS_HASH_KEY to getCurrentMgmGroupParametersHash()
         )
         return TransactionMetadataImpl(metadata + ledgerSpecificMetadata)
     }
 
+    private fun getCurrentMgmGroupParametersHash(): String {
```
Comment (kyriathar) [inline components/ledger/ledger-common-flow/src/main/kotlin/net/corda/ledger/common/flow/impl/transaction/factory/TransactionMetadataFactoryImpl.kt]: Not familiar with group parameters but I believe this should be renamed to `getCurrentMgmGroupParametersId` and maybe `InternalGroupParameters.hash` to id too, hash is an implementation detail?

Comment (vlajos) [reply components/ledger/ledger-common-flow/src/main/kotlin/net/corda/ledger/common/flow/impl/transaction/factory/TransactionMetadataFactoryImpl.kt]: I think they use `epoch` as a primary key.
`Hash` is the hash of the group parameters key-value store's binary representation what the ledger use as an identifier.
It is easy to update, but I think it reflects quite well that it is also a "fingerprint" of the the parameters.

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/UtxoRepository.kt [hunk 2]
```diff
@@ -149,4 +150,21 @@ interface UtxoRepository {
...
+        entityManager: EntityManager,
+        hash: String
+    ): SignedGroupParameters?
+
+    /** Persists signed group parameters */
+    @Suppress("LongParameterList")
+    fun persistSignedGroupParameters(
+        entityManager: EntityManager,
+        hash: String,
+        parameters: ByteArray,
```
Comment (blsemo) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/UtxoRepository.kt]: I don't like the asymmetry here - your persist the serialized group paramters as `ByteArray`, but when you load it, you get back the deserialized object.

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoRepositoryImpl.kt [hunk 3]
```diff
@@ -441,6 +445,58 @@ class UtxoRepositoryImpl @Activate constructor(
...
+                SignedGroupParameters(
+                    ByteBuffer.wrap(r.get(0) as ByteArray),
+                    CryptoSignatureWithKey(
+                        ByteBuffer.wrap(r.get(1) as ByteArray),
+                        ByteBuffer.wrap(r.get(2) as ByteArray)
+                    ),
+                    CryptoSignatureSpec((r.get(3) as String), null, null)
+                )
+            }
+            .firstOrNull()
```
Comment (blsemo) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoRepositoryImpl.kt]: Should that be `singleOrNull()`? If the hash is the primary key, than finding 2 would be definite error?

File: testing/ledger/ledger-utxo-base-test/src/main/kotlin/net/corda/ledger/utxo/test/UtxoLedgerTest.kt [hunk 4]
```diff
@@ -64,11 +67,13 @@ abstract class UtxoLedgerTest : CommonLedgerTest() {
...
         serializationServiceNullCfg,
-        mockTransactionSignatureService(),
+        fakeTransactionSignatureService(),
         transactionMetadataFactory,
         wireTransactionFactory,
         utxoLedgerTransactionFactory,
-        mockUtxoLedgerTransactionVerificationService
+        mockUtxoLedgerTransactionVerificationService,
+        mockUtxoLedgerGroupParametersPersistenceService,
+        mockMembershipGroupReaderProvider
```
Comment (blsemo) [inline testing/ledger/ledger-utxo-base-test/src/main/kotlin/net/corda/ledger/utxo/test/UtxoLedgerTest.kt]: Why did you rename some of the mock services to fake services? Is there a fundamental difference between fakes and mocks here?

**Review by blsemo — COMMENTED** [review]
I am getting confused between `PersistSignedGroupParametersIfDoNotExistParameters`, `SignedGroupParameters`, `CordaSignedGroupParamateres` and `SignedGroupParamtersContainer` - can we simplify that somehow? It feels

File: testing/ledger/ledger-utxo-base-test/src/main/kotlin/net/corda/ledger/utxo/test/UtxoLedgerTest.kt (hunk 4 above)
Comment (vlajos) [reply testing/ledger/ledger-utxo-base-test/src/main/kotlin/net/corda/ledger/utxo/test/UtxoLedgerTest.kt]: I think they are closer to fakes than to mocks.
Difference is not too much.
We can say naming hygiene.

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoRepositoryImpl.kt (hunk 3 above)
Comment (vlajos) [reply components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoRepositoryImpl.kt]: Thanks. I have updated it.

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/UtxoRepository.kt (hunk 2 above)
Comment (vlajos) [reply components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/UtxoRepository.kt]: Thanks. I have updated it to use the object in persist too.

**Comment by vlajos** [comment]
> I am getting confused between `PersistSignedGroupParametersIfDoNotExistParameters`, `SignedGroupParameters`, `CordaSignedGroupParamateres` and `SignedGroupParamtersContainer` - can we simplify that somehow? It feels

Yeah. It was quite confusing.
I have removed the container and added a serializer for the SignedGroupParameters to make it usable in flows.
The other parts are the standard persistence patterns.

File: components/ledger/ledger-persistence/src/integrationTest/kotlin/net/corda/ledger/persistence/consensual/tests/ConsensualLedgerMessageProcessorTests.kt [hunk 5]
```diff
@@ -169,16 +176,26 @@ class ConsensualLedgerMessageProcessorTests {
         assertThat(response.error).isNull()
         val entityResponse = deserializer.deserialize(response.payload.array())!!
         assertThat(entityResponse.results).hasSize(1)
-        assertThat(entityResponse.results.first()).isEqualTo(serializedTransaction)
```
Comment (vlajos) [inline components/ledger/ledger-persistence/src/integrationTest/kotlin/net/corda/ledger/persistence/consensual/tests/ConsensualLedgerMessageProcessorTests.kt:172]: It is not guaranteed that these serialized forms are the same.
It seems that having multiple empty lists in the component groups generates a different result depending on whether they are the same instances or not.

Comment (mnesbit) [reply components/ledger/ledger-persistence/src/integrationTest/kotlin/net/corda/ledger/persistence/consensual/tests/ConsensualLedgerMessageProcessorTests.kt:172]: Ah! Have we copied wrong AMQP serialization mode! In C4 there are differences between storage and messaging AMQP profiles. In particular 'storage', which is used for transactions and things that need to be consistently signed doesn't tokenise repeated object instances. It will treat them as separate items and serialize their fields all over again. Obviously in a complex object graph this can massively expand the data size, but it is only way to get a consistent serialized form.

Comment (vlajos) [reply components/ledger/ledger-persistence/src/integrationTest/kotlin/net/corda/ledger/persistence/consensual/tests/ConsensualLedgerMessageProcessorTests.kt:172]: I added some more details on slack: https://r3holdco.slack.com/archives/C03A3TQMBHR/p1681295411889109

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/persistence/UtxoLedgerGroupParametersPersistenceServiceImpl.kt [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+                requireNotNull(it.signature){
+                    "Group parameters need to be signed."
+                }
+                requireNotNull(it.signatureSpec){
+                    "Group parameters signature need a signature specification."
+                }
+                it
+            }
+        }
+    }
```
Comment (lankydan) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/persistence/UtxoLedgerGroupParametersPersistenceServiceImpl.kt]:
```kotlin
        }.singleOrNull()?.also { groupParameters ->
                requireNotNull(groupParameters.signature){
                    "Group parameters need to be signed."
                }
                requireNotNull(groupParameters.signatureSpec){
                    "Group parameters signature need a signature specification."
                }
        }
```

Add the variable name because its not clear what it is, I assume its `groupParameters` but I don't actually know, hence comment.

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/persistence/external/events/FindSignedGroupParametersExternalEventFactory.kt [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+                .build()
+        )
+    }
+
+    override fun resumeWith(checkpoint: FlowCheckpoint, response: FindSignedGroupParametersResponse): List<SignedGroupParameters> {
+        return response.results.map {
+            requireNotNull(it.mgmSignature) {
+                "Group parameters need to be signed."
+            }
+            requireNotNull(it.mgmSignatureSpec) {
```
Comment (lankydan) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/persistence/external/events/FindSignedGroupParametersExternalEventFactory.kt:59]: The avro one is `mgmSignatureSpec` but the non avro one is `signatureSpec`, do we want to align these?

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/serializer/kryo/SignedGroupParametersKryoSerializer.kt [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+import org.osgi.service.component.annotations.Component
+import org.osgi.service.component.annotations.Reference
+import org.osgi.service.component.annotations.ServiceScope
+
+@Component(
+    service = [ CheckpointInternalCustomSerializer::class, UsedByFlow::class ],
+    property = [SandboxConstants.CORDA_UNINJECTABLE_SERVICE],
+    scope = ServiceScope.PROTOTYPE
+)
+class SignedGroupParametersKryoSerializer @Activate constructor(
```
Comment (lankydan) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/serializer/kryo/SignedGroupParametersKryoSerializer.kt:22]: Not saying its wrong, but why do we need a custom serializer for this class?

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/factory/impl/UtxoSignedTransactionFactoryImpl.kt [hunk 9]
```diff
@@ -96,13 +111,38 @@ class UtxoSignedTransactionFactoryImpl @Activate constructor(
...
+        val groupReader = membershipGroupReaderProvider.getGroupReader(holdingIdentity)
+        val signedGroupParameters = requireNotNull(groupReader.signedGroupParameters) {
+            "Group parameters could not be accessed."
+        }
+        requireNotNull(signedGroupParameters.signature) {
+            "Group parameters need to be signed."
+        }
+        requireNotNull(signedGroupParameters.signatureSpec) {
+            "Group parameters signature need a signature specification."
+        }
```
Comment (lankydan) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/factory/impl/UtxoSignedTransactionFactoryImpl.kt]: Seen these 2 requires quite a few time through the code, should we consolidate them? Happy to push to tech debt if needed, but it will probably never be done if we do that.

File: libs/ledger/ledger-common-data/src/main/kotlin/net/corda/ledger/common/data/transaction/factory/WireTransactionFactoryImpl.kt [hunk 10]
```diff
@@ -42,6 +43,12 @@ class WireTransactionFactoryImpl @Activate constructor(
     }
 
     override fun create(componentGroupLists: List<List<ByteArray>>): WireTransaction {
+        val metadata =
+            parseMetadata(componentGroupLists[TransactionMetadataImpl.ALL_LEDGER_METADATA_COMPONENT_GROUP_ID].first())
+        check((metadata as TransactionMetadataInternal).getNumberOfComponentGroups() == componentGroupLists.size){
+            "Number of component groups in metadata structure description does not match with the real number!" +
+            (metadata as TransactionMetadataInternal).getNumberOfComponentGroups() + " vs " + componentGroupLists.size
+        }
```
Comment (lankydan) [inline libs/ledger/ledger-common-data/src/main/kotlin/net/corda/ledger/common/data/transaction/factory/WireTransactionFactoryImpl.kt:51]: Seen this check a number of times as well.

File: testing/ledger/ledger-common-testkit/src/main/kotlin/net/corda/ledger/common/testkit/WireTransactionExample.kt [hunk 11]
```diff
@@ -10,25 +10,30 @@ import net.corda.v5.application.marshalling.JsonMarshallingService
...
-            ledgerModel = ledgerModel,
-            transactionSubType = transactionSubType
-        )
     val metadataJson = jsonMarshallingService.format(metadata)
     val canonicalJson = jsonValidator.canonicalize(metadataJson)
 
     val allGroupLists = listOf(
         listOf(canonicalJson.toByteArray()),
-    ) + componentGroups
+    ) +
```
Comment (lankydan) [inline testing/ledger/ledger-common-testkit/src/main/kotlin/net/corda/ledger/common/testkit/WireTransactionExample.kt]: This formatting is ugly.

File: testing/ledger/ledger-utxo-base-test/src/main/kotlin/net/corda/ledger/utxo/test/MockMemberShipGroupReaderProvider.kt [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+import org.mockito.Mockito
+import org.mockito.kotlin.any
+
+internal fun mockMembershipGroupReaderProvider(): MembershipGroupReaderProvider {
+
+    val mockMembershipGroupReaderProvider = Mockito.mock(MembershipGroupReaderProvider::class.java)
+    val mockGroupReader = Mockito.mock(MembershipGroupReader::class.java)
+    val mockSignedGroupParameters = Mockito.mock(SignedGroupParameters::class.java)
+
+    Mockito.`when`(mockSignedGroupParameters.bytes).thenReturn(byteArrayOf(1, 1, 11))
```
Comment (lankydan) [inline testing/ledger/ledger-utxo-base-test/src/main/kotlin/net/corda/ledger/utxo/test/MockMemberShipGroupReaderProvider.kt]: Can't we use `whenever` instead of `Mockito.when`? Or is this part of the OSGi tests? If so then ignore my comment.

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/serializer/kryo/SignedGroupParametersKryoSerializer.kt (hunk 8 above)
Comment (vlajos) [reply components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/serializer/kryo/SignedGroupParametersKryoSerializer.kt:22]: SignedGroupParameters have an internal lambda, which is not too Kryo friendly...
https://r3holdco.slack.com/archives/C03A3TQMBHR/p1681253250657469

File: testing/ledger/ledger-utxo-base-test/src/main/kotlin/net/corda/ledger/utxo/test/MockMemberShipGroupReaderProvider.kt (hunk 12 above)
Comment (vlajos) [reply testing/ledger/ledger-utxo-base-test/src/main/kotlin/net/corda/ledger/utxo/test/MockMemberShipGroupReaderProvider.kt]: If I remember right, these ones are used in OSGi tests.

File: testing/ledger/ledger-common-testkit/src/main/kotlin/net/corda/ledger/common/testkit/WireTransactionExample.kt (hunk 11 above)
Comment (vlajos) [reply testing/ledger/ledger-common-testkit/src/main/kotlin/net/corda/ledger/common/testkit/WireTransactionExample.kt]: I have reformatted it a bit.

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/persistence/UtxoLedgerGroupParametersPersistenceServiceImpl.kt (hunk 6 above)
Comment (vlajos) [reply components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/persistence/UtxoLedgerGroupParametersPersistenceServiceImpl.kt]: I have updated the messages.

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/persistence/external/events/FindSignedGroupParametersExternalEventFactory.kt (hunk 7 above)
Comment (vlajos) [reply components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/persistence/external/events/FindSignedGroupParametersExternalEventFactory.kt:59]: It sounds a good idea. I'll open a tech debt ticket for it.

File: processors/flow-processor/build.gradle [hunk 13]
```diff
@@ -50,6 +50,7 @@ dependencies {
     runtimeOnly project(':libs:ledger:ledger-consensual-data')
     runtimeOnly project(':components:ledger:ledger-utxo-flow')
     runtimeOnly project(':libs:ledger:ledger-utxo-data')
+    runtimeOnly project(':libs:ledger:ledger-utxo-transaction-verifier') // todo revert with CORE-8956
```
Comment (lankydan) [inline processors/flow-processor/build.gradle]: Why are new processors needed?

Comment (vlajos) [reply processors/flow-processor/build.gradle]: Group policy provider needs membershipQueryClient and membershipPersistenceClient too.
The group policy provider is needed for the MGM public key what we use the confirm the authenticity of the signed group parameters.

Comment (vlajos) [reply processors/flow-processor/build.gradle]: (ledger-utxo-transaction-verifier was just my first guess. That one was not needed here.)

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/persistence/external/events/FindSignedGroupParametersExternalEventFactory.kt (hunk 7 above)
Comment (vlajos) [reply components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/persistence/external/events/FindSignedGroupParametersExternalEventFactory.kt:59]: https://r3-cev.atlassian.net/browse/CORE-12457

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoPersistenceServiceImpl.kt [hunk 14]
```diff
@@ -217,4 +222,24 @@ class UtxoPersistenceServiceImpl constructor(
...
+
+    override fun findSignedGroupParameters(hash: String): SignedGroupParameters? {
+        return entityManagerFactory.transaction { em ->
+            repository.findSignedGroupParameters(em, hash)
+        }
+    }
+
+    override fun persistSignedGroupParametersIfDoNotExist(signedGroupParameters: SignedGroupParameters) {
+        val cordaSignedGroupParameters = groupParametersFactory.create(signedGroupParameters) as CordaSignedGroupParameters
+        val hash = cordaSignedGroupParameters.hash.toString()
```
Comment (blsemo) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoPersistenceServiceImpl.kt:275]: Do we need to create this object just to obtain the hash? Could we make the hashing external so we don't need a roundtrip through the factory?

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/factory/impl/UtxoSignedTransactionFactoryImpl.kt (hunk 9 above)
Comment (vlajos) [reply components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/factory/impl/UtxoSignedTransactionFactoryImpl.kt]: I could remove two of these checks. (They are not needed since I converted back the code to use the non-avro SignedGroupParameters.)

File: libs/ledger/ledger-common-data/src/main/kotlin/net/corda/ledger/common/data/transaction/factory/WireTransactionFactoryImpl.kt (hunk 10 above)
Comment (vlajos) [reply libs/ledger/ledger-common-data/src/main/kotlin/net/corda/ledger/common/data/transaction/factory/WireTransactionFactoryImpl.kt:51]: I removed two of these as well from the consensual/utxo signed tx factory. (The same check happened a few lines before in wire tx factory.)

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoPersistenceServiceImpl.kt (hunk 14 above)
Comment (vlajos) [reply components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoPersistenceServiceImpl.kt:275]: That sounds like bypassing the public API and depending on the implementation.
It is probably quicker, but sounds less clean for me.

Comment (vlajos) [reply components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoPersistenceServiceImpl.kt:275]: I tried it. It looks better than I had expected.
I created another PR for it:
https://github.com/corda/corda-runtime-os/pull/3615
(just to keep this one green.)

**Review by charlieR3 — APPROVED** [review]
@corda/corda-platform-network-team  related changes LGTM.

**Review by lankydan — APPROVED** [review]

**Review by seanbrereton — APPROVED** [review]
LGTM - gradle only
